client/my-sites/controller.js

```diff
@@ -462,48 +462,59 @@ export function siteSelection( context, next ) {
 		return next();
 	}
 
-	const siteId = getSiteId( getState(), siteFragment );
-	if ( siteId ) {
-		// onSelectedSiteAvailable might render an error page about domain-only sites or redirect
-		// to wp-admin. In that case, don't continue handling the route.
-		dispatch( setSelectedSiteId( siteId ) );
-		if ( onSelectedSiteAvailable( context ) ) {
-			next();
+	Promise.resolve( () => {
+		// For unlinked checkout flow, we should always request fresh site information.
+		// Because sites might not be current.
+		if ( '1' !== context?.query?.unlinked || ! context?.path?.startsWith( '/checkout/' ) ) {
+			return;
 		}
-	} else {
-		// Fetch the site by siteFragment and then try to select again
-		dispatch( requestSite( siteFragment ) )
-			.catch( () => null )
-			.then( ( site ) => {
-				let freshSiteId = getSiteId( getState(), siteFragment );
-
-				if ( ! freshSiteId ) {
-					const wpcomStagingFragment = siteFragment.replace(
-						/\b.wordpress.com/,
-						'.wpcomstaging.com'
-					);
-					freshSiteId = getSiteId( getState(), wpcomStagingFragment );
-				}
-
-				if ( freshSiteId ) {
-					// onSelectedSiteAvailable might render an error page about domain-only sites or redirect
-					// to wp-admin. In that case, don't continue handling the route.
-					dispatch( setSelectedSiteId( freshSiteId ) );
-					if ( onSelectedSiteAvailable( context ) ) {
-						next();
-					}
-				} else if ( shouldRedirectToJetpackAuthorize( context, site ) ) {
-					navigate( getJetpackAuthorizeURL( context, site ) );
-				} else {
-					// If the site has loaded but siteId is still invalid then redirect to allSitesPath.
-					const allSitesPath = addQueryArgs(
-						{ site: siteFragment },
-						sectionify( context.path, siteFragment )
-					);
-					page.redirect( allSitesPath );
+		return dispatch( requestSites() );
```

Comment:
The issue is caused by when the site is disconnected and user wouldn't have permission at all while the site is still in his site list. Does that make sense? The site needs to be removed from his site list. So that's where requestSites comes in.